File: packages/cloudinary/android/src/main/java/io/capawesome/capacitorjs/plugins/cloudinary/CloudinaryPlugin.java
package io.capawesome.capacitorjs.plugins.cloudinary;

import androidx.annotation.NonNull;
import com.getcapacitor.Bridge;
import com.getcapacitor.JSObject;
import com.getcapacitor.Logger;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginHandle;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import java.util.HashMap;
import java.util.Map;

@CapacitorPlugin(name = "Cloudinary")
public class CloudinaryPlugin extends Plugin {

    public static final String TAG = "Cloudinary";
    public static final String ERROR_NOT_INITIALIZED = "Plugin is not initialized.";
    public static final String ERROR_CLOUD_NAME_MISSING = "cloudName must be provided.";
    public static final String ERROR_PATH_MISSING = "path must be provided.";
    public static final String ERROR_RESOURCE_TYPE_MISSING = "resourceType must be provided.";
    public static final String ERROR_UPLOAD_PRESET_MISSING = "uploadPreset must be provided.";
    public static final String ERROR_URL_MISSING = "url must be provided.";
    public static Bridge staticBridge = null;

    private Cloudinary implementation;
    private boolean initialized = false;

    public void load() {
        staticBridge = this.bridge;
        implementation = new Cloudinary(this);
    }

    @PluginMethod
    public void initialize(PluginCall call) {
        try {
            if (initialized) {
                return;
            }
            String cloudName = call.getString("cloudName");
            if (cloudName == null) {
                call.reject(ERROR_CLOUD_NAME_MISSING);
                return;
            }

            implementation.initialize(cloudName);
            initialized = true;
            call.resolve();
        } catch (Exception exception) {
            call.reject(exception.getMessage());
            Logger.error(TAG, exception.getMessage(), exception);
        }
    }

    @PluginMethod
    public void uploadResource(PluginCall call) {
        try {
            if (!initialized) {
                call.reject(ERROR_NOT_INITIALIZED);
                return;
            }
            String resourceType = call.getString("resourceType");
            if (resourceType == null) {
                call.reject(ERROR_RESOURCE_TYPE_MISSING);
                return;
            }
            String uploadPreset = call.getString("uploadPreset");
            if (uploadPreset == null) {
                call.reject(ERROR_UPLOAD_PRESET_MISSING);
                return;
            }
            String path = call.getString("path");
            if (path == null) {
                call.reject(ERROR_PATH_MISSING);
                return;
            }
            String publicId = call.getString("publicId");

            implementation.uploadResource(
                resourceType,
                path,
                uploadPreset,
                publicId,
                new UploadResourceResultCallback() {
                    @Override
                    public void success(Map resultData) {
                        JSObject result = CloudinaryHelper.createUploadResourceResult(resultData);
                        call.resolve(result);
                    }

                    @Override
                    public void error(String message) {
                        call.reject(message);
                    }
                }
            );
        } catch (Exception exception) {
            call.reject(exception.getMessage());
            Logger.error(TAG, exception.getMessage(), exception);
        }
    }

    @PluginMethod
    public void downloadResource(PluginCall call) {
        try {
            if (!initialized) {
                call.reject(ERROR_NOT_INITIALIZED);
                return;
            }
            String url = call.getString("url");
            if (url == null) {
                call.reject(ERROR_URL_MISSING);
                return;
            }

            implementation.downloadResource(
                url,
                new DownloadResourceResultCallback() {
                    @Override
                    public void success(String path) {
                        JSObject result = new JSObject();
                        result.put("path", path);
                        call.resolve(result);
                    }

                    @Override
                    public void error(String message) {
                        call.reject(message);
                    }
                }
            );
        } catch (Exception exception) {
            call.reject(exception.getMessage());
            Logger.error(TAG, exception.getMessage(), exception);
        }
    }

    public Cloudinary getImplementation() {
        return implementation;
    }

    public static void onDownloadCompleted(long downloadId) {
        CloudinaryPlugin plugin = CloudinaryPlugin.getCloudinaryPluginInstance();
        if (plugin != null) {
            plugin.getImplementation().handleDownloadCompleted(downloadId);
        }
    }

    private static CloudinaryPlugin getCloudinaryPluginInstance() {
        if (staticBridge == null || staticBridge.getWebView() == null) {
            return null;
        }
        PluginHandle handle = staticBridge.getPlugin("Cloudinary");
        if (handle == null) {
            return null;
        }
        return (CloudinaryPlugin) handle.getInstance();
    }
}

File: packages/cloudinary/ios/Plugin/CloudinaryPlugin.swift
import Foundation
import Capacitor

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(CloudinaryPlugin)
public class CloudinaryPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "CloudinaryPlugin"
    public let jsName = "Cloudinary"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "initialize", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "uploadResource", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "downloadResource", returnType: CAPPluginReturnPromise)
    ]
    public let errorNotInitialized = "Plugin is not initialized."
    public let errorFileNotFound = "The file was not found."
    public let errorUnknown = "An unknown error occurred."
    public let errorCloudNameMissing = "cloudName must be provided."
    public let errorPathMissing = "path must be provided."
    public let errorResourceTypeMissing = "resourceType must be provided."
    public let errorUploadPresetMissing = "uploadPreset must be provided."
    public let errorUrlMissing = "url must be provided."

    private var implementation: Cloudinary?
    private var initialized = false

    override public func load() {
        implementation = Cloudinary(plugin: self)
    }

    @objc func initialize(_ call: CAPPluginCall) {
        guard let cloudName = call.getString("cloudName") else {
            call.reject(errorCloudNameMissing)
            return
        }
        implementation?.initialize(cloudName)
        initialized = true
        call.resolve()
    }

    @objc func uploadResource(_ call: CAPPluginCall) {
        if !initialized {
            call.reject(errorNotInitialized)
            return
        }
        guard let resourceType = call.getString("resourceType") else {
            call.reject(errorResourceTypeMissing)
            return
        }
        guard let uploadPreset = call.getString("uploadPreset") else {
            call.reject(errorUploadPresetMissing)
            return
        }
        guard let path = call.getString("path") else {
            call.reject(errorPathMissing)
            return
        }
        let publicId = call.getString("publicId")
        implementation?.uploadResource(resourceType: resourceType, path: path, uploadPreset: uploadPreset, publicId: publicId, completion: { resultData, errorMessage in
            if let errorMessage = errorMessage {
                call.reject(errorMessage)
                return
            }
            guard let resultData = resultData else {
                call.reject(self.errorUnknown)
                return
            }
            let result = CloudinaryHelper.createUploadResourceResult(resultData)
            call.resolve(result)
        })
    }

    @objc func downloadResource(_ call: CAPPluginCall) {
        if !initialized {
            call.reject(errorNotInitialized)
            return
        }
        guard let url = call.getString("url") else {
            call.reject(errorUrlMissing)
            return
        }
        implementation?.downloadResource(url: url, completion: { path, errorMessage in
            if let errorMessage = errorMessage {
                call.reject(errorMessage)
                return
            }
            guard let path = path else {
                call.reject(self.errorUnknown)
                return
            }
            call.resolve([
                "path": path
            ])
        })
    }
}
